### PR TITLE
EE-1120: Feat/registry sync

### PR DESCRIPTION
**Pull request by SvenDowideit**
# Replaced by https://github.com/portainer/agent/pull/250
https://portainer.atlassian.net/browse/EE-1120

just the last commit - the rest if the refactoring from #229

### CODE REVIEW

File: edge/key.go [hunk 1]
```diff
@@ -122,3 +134,71 @@ func encodeKey(edgeKey *edgeKey) string {
...
+
+	return edgeKey, nil
+}
+
+func retrieveEdgeKeyFromFilesystem() (string, error) {
+	var edgeKey string
+
+	edgeKeyFilePath := fmt.Sprintf("%s/%s", agent.DataDirectory, agent.EdgeKeyFile)
+
+	keyFileExists, err := filesystem.FileExists(edgeKeyFilePath)
```
Comment (andres-portainer) [inline edge/key.go:167]: Nitpick: this check seems a bit redundant (and not concurrency-friendly), if the file doesn't exist then `filesystem.ReadFromFile()` will catch it below anyway.

File: edge/key.go [hunk 2]
```diff
@@ -20,15 +21,26 @@ type edgeKey struct {
...
 func (manager *Manager) SetKey(key string) error {
 	edgeKey, err := parseEdgeKey(key)
 	if err != nil {
 		return err
 	}
 
-	err = filesystem.WriteFile(agent.DataDirectory, agent.EdgeKeyFile, []byte(key), 0444)
-	if err != nil {
-		return err
+	// TODO: yeah, we don't know if we got it from the fs, so lets try again :/
```
Comment (andres-portainer) [inline edge/key.go:31]: Maybe just write the file with `os.O_CREATE | os.O_EXCL` instead?

File: edge/key.go (hunk 1 above)
Comment (deviantony) [reply edge/key.go:167]: Referenced in https://github.com/portainer/agent/issues/248

File: edge/key.go (hunk 2 above)
Comment (deviantony) [reply edge/key.go:31]: Got a reference for that one in https://github.com/portainer/agent/pull/239

**Comment by cheloRydel** [comment]
Replaced by https://github.com/portainer/agent/pull/250

**Pull request closed without merging**